File: chap5_DFS_BFS/codetree_glacier.py
'''
2023.09.14
'''
# 처음 작성하던 답


from collections import deque
n, m = map(int, input().split())

graph = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [
    [False for _ in range(m)]
    for _ in range(n)
]
water_visited = [
    [False for _ in range(m)]
    for _ in range(n)
]
is_surrounded_water = [
    [False for _ in range(m)]
    for _ in range(n)
]
time = 0 # 빙하가 전부 녹는데 걸리는 시간
melt_cnt = 0 # 녹은 빙하의 크기, 맨 마지막 것을 답으로
meltable_list = [] # 매 초마다 초기화 하면서 각 시간별 녹일 수 있는 빙하 좌표 담기

dx = [-1,1,0,0]
dy = [0,0,-1,1]

#### 주목, 이제 check_water()만 하면 이 문제도 끝나는거야!!!!!
def check_water():
    # 이 함수는 0이 1로 둘러싸였는지 아닌지 체크
    # 만약 0이 1로 둘러싸였으면?--> is_surrounded_water를 True

    # step1: 연결요소 먼저 찾고
    # step2: 걔를 둘러싼 격자칸이 1인지 확인
    # 만약 둘러싼 격자칸이 1이다? 그럼 그 연결요소 모두 is_surrounded_water = True
    for i in range(1,n-1):
        for j in range(1, m-1):
            if graph[i][j] == 0

def bfs(x,y):
    global meltable_list
    meltable_list = []
    q = deque()
    visited[x][y] = True
    q.append((x,y))

    while q:
        a, b = q.popleft()
        for i in range(4):
            nx = a + dx[i]
            ny = b + dy[i]

            if nx < 1 or nx >= n-1 or ny < 1 or ny >= m-1:
                continue
            if visited[nx][ny]:
                continue
            if not graph[nx][ny]:
                continue


            # 여기서 중요한 건 상하좌우에 현재 탐색하는 1과 맞닿아 있는 0이 있는지
            
            # 그럼 여기서 또 dx dy 써야하는건가?
            # 현재 1인 좌표의 상하좌우 탐색함
            valid_stone = 0
            for i in range(4):
                nx_s = nx + dx[i]
                ny_s = ny + dy[i]
                if not graph[nx_s][ny_s] and not is_surrounded_water[nx_s][ny_s]:# 해당 돌이 1로 둘러싸이지 않았다면
                    valid_stone += 1
            # 그 맞닿아 있는 0이 있고, 그 0 중 하나라도 1로 둘러싸이지 않았다면 meltable_list에 넣기
            if valid_stone > 0:
                meltable_list.append((nx, ny))
            # 그리고 큐에 넣기
            visited[nx][ny] = True
            q.append((nx,ny))



# 모든 빙하가 녹을 때까지 반복
while 1:
    remain_ice = sum(graph, []).count(1)
    if remain_ice == 0:
        break
    melt_cnt = 0
    
    # 먼저 1로 둘러싸인 물이 있는지 확인
    for i in range(1,n-1):
        for j in range(1,m-1):
            if not graph[i][j] and not water_visited[i][j]:
        
    check_water()
    for i in range(1,n-1):
        for j in range(1, m-1):
            if graph[i][j] and not visited[i][j]:
                bfs(i,j) # bfs 탐색 돌기, 돌면서 확인하는데

    # 녹일 빙하 0으로 만들기/ 녹일 빙하 크기 melt_cnt에 담기
    
    for ice in meltable_list:
        a,b = ice
        graph[a][b] = 0
        melt_cnt += 1
    time += 1

print(time, melt_cnt)


## 최종 답
import sys
input = sys.stdin.readline
from collections import deque

n, m = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(n)]

dx = [-1,1,0,0]
dy = [0,0,-1,1]

'''
(0,0) 에서 시작해서 BFS 수행하면 외부에서 이어져 있는 물이니 빙하로 둘러싸인 
'''
# 격자 가장 바깥 부분의 외부 물부터 시작해서 상하좌우로 이동, 만나는 빙하를 녹임
# BFS를 계속 수행하면, 빙하 바깥쪽의 물에서 시작하여 빙하 내부의 물까지도 방문할 수 있음
# 빙하 내부의 물에 도달하면, 그 물은 빙하에 둘러싸인 물이 아니게 됨, 빙하로 막혀있다면 BFS를 통해 접근할 수 없었을 것이기 때문

def bfs():
    global graph
    # 매 초마다 새로운 graph에 대해 bfs 수행하니 visited도 초기화 시켜줌
    visited = [[False for _ in range(m)] for _ in range(n)]
    next_graph = [row.copy() for row in graph]
    
    q = deque()
    
    # 빙하에 둘러싸이지 않는 물 탐색
    q.append((0,0))
    visited[0][0] = True
    
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny]:
                visited[nx][ny] = True
                if graph[nx][ny] == 1: # 만약 물과 맞닿아 있는 빙하라면
                    next_graph[nx][ny] = 0 # 녹이기
                else: # 만약 외부 물과 연결된 물이라면
                    q.append((nx,ny)) # 큐에 넣어서 이후에 맞닿아 있는 빙하 도달할 수 있도록 함
    return next_graph # 빙하가 녹은 후의 상태 반환

time = 0
while sum(map(sum,graph)) > 0 : # 빙하가 완전히 녹을 떄까지 (합이 0이 되면 break)
    melt_cnt = sum(map(sum,graph)) # 현재 빙하의 양
    graph = bfs() # 빙하 녹이고 나온 결과 
    time += 1 # 시간 증가

print(time, melt_cnt)

## 해설
'''
빙하가 녹을 조건: 인접한 곳에 빙하에 둘러싸여 있지 않은 물이 있는 경우
(0,0)을 시작점으로 BFS 돌리면 빙하에 둘러싸이지 않은 물의 영역 구할 수 있음
'''

from collections import deque
import enum

class Element(enum.Enum):
    WATER = 0
    GLACIER = 1
    

# 변수 선언 및 입력
n, m = tuple(map(int, input().split()))

a = [
    list(map(int, input().split()))
    for _ in range(n)
]

# bfs에 필요한 변수들
q = deque()
glaciers_to_melt = deque()

visited = [
    [False for _ in range(m)]
    for _ in range(n)
]

cnt = 0

# 0: 오른쪽, 1: 아래쪽, 2: 왼쪽, 3: 위쪽
dxs, dys = [0,1,0,-1], [1,0,-1,0]

# 소요 시간과 가장 마지막으로 녹은 빙하의 수를 저장
elapsed_time = 0
last_melt_cnt = 0

# 주어진 위치가 격자를 벗어나는지 여부를 반환
def in_range(x,y):
    return 0 <= x < n and 0 <= y < m

# 범위를 벗어나지 않으면서 물이어야 하고 방문한 적이 없어야 갈 수 있음
def can_go(x,y):
    return in_range(x,y) and a[x][y] == Element.WATER.value and not visited[x][y]

def is_glacier(x,y):
    return in_range(x,y) and a[x][y] == Element.GLACIER.value and not visited[x][y]

# visited 배열을 초기화
def initialize():
    for i in range(n):
        for j in range(m):
            visited[i][j] = False
            
# 빙하에 둘러싸여 있지 않은 물을 전부 구해주는 BFS
# 문제에서 가장자리는 전부 물로 주어짐
# 따라서 항상 (0,0)에서 시작하여 탐색을 진행하면
# 빙하에 둘러싸여 있지 않은 물들은 전부 visited 처리가 됨

def bfs():
    # BFS 함수가 여러 번 호출되므로
    # 사용하기 전 visited 배열을 초기화해줌
    initialize()
    
    # 항상 (0,0)에서 시작
    q.append((0,0))
    visited[0][0] = True
    
    while q:
        # queue에서 가장 먼저 들어온 원소를 뺌
        x, y = q.popleft()
        
        # queue에서 뺸 원소의 위치를 기준으로 네 방향을 확인
        for dx, dy in zip(dxs, dys):
            new_x, new_y = x + dx, y + dy
            
            
            # 더 갈 수 있는 곳이면 queue에 추가
            if can_go(new_x, new_y):
                q.append((new_x, new_y))
                visited[new_x][new_y] = True
                
    
# 현재 위치를 기준으로 인접한 영역에
# 빙하에 둘러싸여 있지 않은 물이 있는지 판단
def outside_water_exist_in_neighbor(x,y):
    for dx, dy in zip(dxs, dys):
        new_x, new_y = x + dx, y + dy
        
        if in_range(new_x, new_y) and visited[new_x][new_y]:
            return True
    return False


# 인접한 영역에 빙하에 둘러싸여 있지 않은 물이 있는 빙하를 찾아 녹여줌
def melt():
    global last_melt_cnt
    
    for i in range(n):
        for j in range(m):
            if a[i][j] == Element.GLACIER.value and \
                outside_water_exist_in_neighbor(i,j):
                a[i][j] = Element.WATER.value
                last_melt_cnt += 1

# 빙하를 한 번 녹임
def simulate():
    global elapsed_time, last_melt_cnt
    
    elapsed_time += 1
    last_melt_cnt = 0
    
    # 빙하에 둘러싸여 있지 않은 물의 위치를 전부 visited 로 체크
    bfs()
    
    # 인접한 영역에 빙하에 둘러싸여 있지 않은 물이 있는 빙하를 찾아 녹임
    melt()
    
# 빙하가 아직 남아 있는지 확인
def glacier_exist():
    for i in range(n):
        for j in range(m):
            if a[i][j] == Element.GLACIER.value:
                return True
    return False

while True:
    simulate()
    
    # 빙하가 존재하는 한 계속 빙하를 녹임
    if not glacier_exist():
        break
print(elapsed_time, last_melt_cnt)

File: chap5_DFS_BFS/codetree_glacier.cpp
/*
 * 2023.1.3
 * 도저히 풀이 방법이 생각나지 않아서 해설을 참고하고 처음부터 다시 풀었음
 * (0,0)에서 탐색 시작 -> 0을 만나면 visited 표시
 * 모든 격자를 돌면서 1이고 옆에 visited표시된 물이 있는지 확인
 * 있으면 0으로 바꾸기
 * 빙하가 다 녹는데 길어야 max(N,M)시간이 걸림, 해당 시간동안 빙하에 둘러싸여 있지 않은 물을 찾기 위해 BFS를 계속 수행해야함: O(N*M*max(N,M)) 시간복잡도
*/

#include <iostream>
#include <queue>

#define MAX_N 200
#define MAX_M 200
#define MAX_DIR 4
using namespace std;
int n, m;
int grid[MAX_N][MAX_M];
queue<pair<int, int> > q;       
bool visited[MAX_N][MAX_M];
int curr_melted;
int dirs[MAX_DIR][2] = {{1,0},{-1,0},{0,1},{0,-1}};
int time_sec;
bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < m;
}

bool CanGo(int x, int y){
    return InRange(x,y) && !visited[x][y] && !grid[x][y]; 
}

void bfs(){
    while(!q.empty()){
        // q가 빌 때까지
        pair<int, int> curr = q.front();
        q.pop();
        for(int i=0; i<MAX_DIR; i++){
            int nx = curr.first + dirs[i][0];
            int ny = curr.second + dirs[i][1];
            if(CanGo(nx,ny)){
                visited[nx][ny] = true;
                q.push(make_pair(nx,ny));
            }
        }

    }
}

bool IsFinish(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            if(grid[i][j]) return false;
        }
    }
    return true;
}

void Initialize(){
    // 1. 방문 배열 초기화
    for(int i=0; i<n; i++)
        for(int j=0; j<m; j++)
            visited[i][j] = false;
    
    // 2. 현재 녹은 빙하 개수 초기화
    curr_melted = 0;
}

void Melt(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            if(grid[i][j]){
                // 만약 빙하라면
                for(int d=0; d<MAX_DIR; d++){
                    int nx = i+dirs[d][0], ny = j+dirs[d][1];
                    if(InRange(nx,ny) && grid[nx][ny] == 0 && visited[nx][ny]){
                        grid[i][j] = 0;
                        curr_melted++;
                        break;
                    }
                }
            }
        }
    }
}

void Simulate(){
    time_sec++;
    // Step 0. 시뮬레이션을 돌리기 전에 먼저 초기화하기
    Initialize();
    // Step 1. 먼저 bfs 탐색을 돌면서 빙하로 둘러싸이지 않은 물을 파악하기
    visited[0][0] = true;
    q.push(make_pair(0,0));
    bfs();

    // Step 2. 모든 격자를 돌면서 녹일 수 있는 빙하는 녹이기
    Melt();

}

int main(){
    cin >> n >> m;
    for(int i=0; i<n; i++)
        for(int j=0; j<m; j++)
            cin >> grid[i][j];
    
    while(true){
        // 모든 격자가 0이 된다면 break;
        if(IsFinish()) break;
        Simulate();
    }
    
    cout << time_sec << ' ' << curr_melted << '\n';
    return 0;
}

/*
 * 해설 풀이 1
*/
#include <iostream>
#include <queue>
#include <algorithm>

enum ELEMENT {
    WATER,
    GLACIER
};

#define DIR_NUM 4
#define MAX_N 200
#define MAX_M 200

using namespace std;

// 전역 변수 선언:
int n, m;

int a[MAX_N][MAX_M];

// bfs에 필요한 변수들
queue<pair<int, int> > q;
bool visited[MAX_N][MAX_M];
int cnt;

int dx[DIR_NUM] = {1,-1,0,0};
int dy[DIR_NUM] = {0,0,1,-1};

// 소요 시간과 가장 마지막으로 녹은 빙하의 수를 저장
int elapsed_time, last_melt_cnt;

// visited 배열을 초기화함
void Initialize(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            visited[i][j] = 0;
        }
    }
}

// 범위가 격자 안에 들어가는지 확인
bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < n;
}

// 범위를 벗어나지 않으면서 물이여야 하고 방문한 적이 없어야 갈 수 있음
bool CanGo(int x, int y){
    return InRange(x,y) && a[x][y] == WATER && !visited[x][y]; 
}

// 빙하에 둘러쌓여 있지 않은 물들을 전부 구해주는 BFS
// 문제에서 가장자리는 전부 물로 주어지기 때문에 항상 (0,0)에서 시작해서 탐색 진행
// 그러면 빙하에 둘러쌓여 있지 않은 물들은 전부 visited 처리가 됨
void BFS(){
    // BFS 함수가 여러번 호출되므로
    // 사용하기 전에 visited 배열을 초기화해줌
    Initialize();

    // 항상 (0,0)에서 시작함
    q.push(make_pair(0,0));
    visited[0][0] = true;

    while(!q.empyt()){
        // queue에서 가장 먼저 들어온 원소를 빼기
        pair<int, int> curr_pos = q.front();
        int x = curr_pos.first, y = curr_pos.second;
        q.pop();

        // queue에서 뺀 원소의 위치를 기준으로 4방향 확인
        for(int dir= 0; dir<DIR_NUM; dir++){
            int nx = x + dx[dir], ny + y + dy[dir];
            

            // 더 갈 수 있는 곳이라면 queue에 추가
            if(CanGo(nx,ny)){
                q.push(make_pair(nx,ny));
                visited[nx][ny] = true;
            }
        }
    }
}

// 현재 위치를 기준으로 인접한 영역에 빙하에 둘러쌓여 있지 않은 물이 있는지 판단하기
bool OutsideWaterExistInNeighbor(int x, int y){
    for(int dir=0; dir<DIR_NUM; dir++){
        int nx = x+dx[dir], ny = y+dy[dir];
        if(InRange(nx,ny) && visited[nx][ny])
            return true;
    }
    return false;
}

// 인접한 영역에 빙하에 둘러쌓여 있지 않은 물이 있는 빙하를 찾아 녹여주기
void Melt(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            if(a[i][j] == GLACIER &&
                OutsideWaterExistInNeighbor(i,j)){
                    a[i][j] = WATER;
                    last_melt_cnt++;
                }
        }
    }

}

// 빙하를 한 번 녹임
void Simulate(){
    elapsed_time++;
    last_melt_cnt = 0;

    // 빙하에 둘러쌓여 있지 않은 물의 위치를 전부 visited로 체크
    BFS();

    // 인접한 영역에 빙하에 둘러쌓여 있지 않은 ㅁ루이 있는 빙하를 찾아 녹여주기
    Melt();
}

// 빙하가 아직 남아있는지 확인
bool GlacierExist(){
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            if(a[i][j] == GLACIER)
                return true;
        }
    }
    return false;
}

int main(){
    // 입력:
    cin >> n >> m;
    for(int i=0; i<n; i++)
        for(int j=0; j<m; j++)
            cin >> a[i][j];
    
    do{
        // 빙하를 한 번 녹입니다.
        Simulate();
    } while(GlacierExist()); // 빙하가 존재하는 한 계속 반복하기

    // 출력:
    cout << elapsed_time << " "  << last_melt_cnt;

    return 0;
}